src/Microsoft.ML.Data/Prediction/Calibrator.cs

```diff
@@ -57,19 +57,19 @@
     "Naive Calibration Executor",
     NaiveCalibrator.LoaderSignature)]
 
-[assembly: LoadableClass(typeof(ValueMapperCalibratedModelParameters<IPredictorProducing<float>, ICalibrator>), null, typeof(SignatureLoadModel),
+[assembly: LoadableClass(typeof(ValueMapperCalibratedModelParameters<object, ICalibrator>), null, typeof(SignatureLoadModel),
```

Comment:
>object [](start = 69, length = 6)

@TomFinley's comment from #2850:
This is a bit stronger medicine than what I was suggesting! (I don't object per se, but it did entail a bit more work.) I suppose it's a bit more "honest" than the old way -- marker interfaces are more or less a promise you can't keep -- but they are still useful on internal code where you have complete control over the implementation (as we do as they are internal).

So I think this should be fine, I just wonder, are we fully comfortable with this?

Reply:
But this is in the public API, since when we load it from a serialized model this would be the type of the object. And then if you want to access `SubModel` (like in the test I added `LoadModelAndExtractPredictor`) you can't do it if it is generic on `IPredictorProducing`.

Another option, is to introduce an abstract class `CalibratedModelParametersBase` that is not generic, where `SubModel` is an object (I made this change to one of the classes in the last commit, so we can decide which option we prefer).

@TomFinley, @eerhardt


---
In reply to: [262684849](https://github.com/dotnet/machinelearning/pull/2858#discussion_r262684849) [](ancestors = 262684849)

Reply:
If we decide to go with the abstract base class, I can do something similar for the `PredictionTransformer`s, except there it would have to be an interface... between `ITransformer` and `IPredictionTransformer<TModel>`.

---
In reply to: [262739008](https://github.com/dotnet/machinelearning/pull/2858#discussion_r262739008) [](ancestors = 262739008,262684849)

Reply:
The other way of course is to just have the type restriction. Then these things become `IPredictionTransformer<object>` via covariance/contravariance (can't remember which this is at the moment). In fact I'd kind of like that even if we go this route.

Reply:
> The other way of course is to just have the type restriction. Then these things become `IPredictionTransformer<object>` via covariance/contravariance (can't remember which this is at the moment). In fact I'd kind of like that even if we go this route, and introduce a "super-interface."

Reply:
I think we kind of needed it in the case you mentioned though, because calibrated models are a class, not an interface, and there's no such thing as these `in`/`out` modifiers on type parameters for classes as there are with interfaces. If we can use that to our advantage to avoid complicating the type hierarchy, I think I'd prefer to.

Reply:
Actually I'm thinking about this now, and I don't like the idea of there being a super-interface, because we already have this hierarchy:

https://github.com/dotnet/machinelearning/blob/adb53c41fd8c612df30f8d0213b7f459ef113724/src/Microsoft.ML.Data/Prediction/IPredictionTransformer.cs#L16

And here:

https://github.com/dotnet/machinelearning/blob/adb53c41fd8c612df30f8d0213b7f459ef113724/src/Microsoft.ML.Data/Prediction/IPredictionTransformer.cs#L27

We could have another interface, another non-generic `IPredictionTransformer` with `object Model { get; }`, and another non-generic `ISingleFeaturePredictionTransformer` that descends from it, with the generic `IPredictionTransformer`. Or we could just make things so that we *know* we can always call them `IPredictionTransformer<object>`.